pallets/poa/src/lib.rs

```diff
@@ -511,41 +762,338 @@ impl<T: Trait> Module<T> {
+                BlockCount::SameBlocks(count) => *count,
             }
-            debug!(
-                target: "runtime",
-                "Epoch {} ends at slot {}",
-                prev_epoch, current_slot_no - 1
+        }
+    }
+
+    /// Return an enum containing either maximum number of blocks produced by any validator if all validators
+    /// did not produce equal blocks in the epoch or the number of blocks if they produced the same number.
+    /// Also return count of blocks produced by each validator in a map.
+    fn count_validator_blocks(
+        current_epoch_no: EpochLen,
+    ) -> (BlockCount, BTreeMap<T::AccountId, EpochLen>) {
+        let mut validator_block_counts = BTreeMap::new();
+        let mut max_blocks = 0;
+        // Flag to check if all validators produced same blocks
+        let mut same_block_count = true;
+        let validators = Self::active_validators();
+        for (i, v) in validators.into_iter().enumerate() {
+            let stats = Self::get_validator_stats_for_epoch(current_epoch_no, &v);
+            let block_count = stats.block_count;
+            if i == 0 {
+                // first iteration
+                max_blocks = block_count
+            } else {
+                if max_blocks != block_count {
+                    // Block count of this validator is different from previous, set the flag as false
+                    same_block_count = false;
+                }
+                if block_count > max_blocks {
+                    max_blocks = block_count;
+                }
+            }
+            debug!(target: "runtime", "Validator {:?} has blocks {}", v, block_count);
+            validator_block_counts.insert(v, block_count);
+        }
+        (
+            if same_block_count {
+                BlockCount::SameBlocks(max_blocks)
+            } else {
+                BlockCount::MaxBlocks(max_blocks)
+            },
+            validator_block_counts,
+        )
+    }
+
+    /// Emission rewards per epoch are set assuming the epoch is of at least `MinEpochLength` and thus
+    /// each validator gets at least `MinEpochLength` / (validator count) number of slots. When this is
+    /// not the case, decrease emission rewards proportionally.
+    /// Assumes that `slots_per_validator` is not 0 and <= expected_slots_per_validator.
+    fn get_max_emission_reward_per_validator_per_epoch(
+        expected_slots_per_validator: EpochLen,
+        slots_per_validator: EpochLen,
+    ) -> u128 {
+        // Maximum emission reward for each validator in an epoch assuming epoch was at least `MinEpochLength`
+        let max_em = Self::max_emm_validator_epoch().saturated_into::<u128>();
+        if slots_per_validator != expected_slots_per_validator {
+            // Reduce the emission for shorter epoch
+            max_em.saturating_mul(slots_per_validator.into())
+                / (expected_slots_per_validator as u128)
+        } else {
+            max_em
+        }
```

Comment:
Wanted to avoid the extra multiplication and division. Might be useless optimization if compiler already does (doubt that it will) that but not sure and did not check the generated code.

Reply:
Your call.                                                                                                                      ^
                                                                                                                                     |
"Commit Suggestion" ---------------------------------------------------------------------------------/

"Resolve Conversation"
                  |
                  v